facilitator/src/transport/s3.rs

```diff
@@ -386,6 +388,17 @@ impl TransportWriter for MultipartUploadWriter {
         // Write last part, if any
         self.upload_part()?;
 
+        if self.completed_parts.is_empty() {
```

Comment:
I'm not so sure about the semantics in this corner case: I think we'd want an upload that has nothing written to it to produce a 0-byte object, rather than return an error.

While users of this functionality may not often want a 0-byte object, if they do there is currently no way to create a 0-byte object. Nor do I imagine a user of this code would ever be able to meaningfully handle the `NoCompletedParts` error. This would also be what happens "naturally" if we were using the non-multipart upload API.

Reply:
For what it's worth, the batch reader does treat missing files as if they encode an empty array of packets. https://github.com/divviup/prio-server/blob/fe45c8d6771ab2ac0a91f9928c1826fcc8d5529c/facilitator/src/batch.rs#L267-L277 On the other hand, I am under the impression workflow-manager will not enqueue a task if the packet file is missing, so that may be moot in practice. https://github.com/divviup/prio-server/blob/fe45c8d6771ab2ac0a91f9928c1826fcc8d5529c/workflow-manager/batchpath/batchpath.go#L173

Reply:
I'm not sure we can use multipart uploads to produce a 0-byte object. I think we're caught in an awkward spot between the behavior of the S3 API and how Rusoto implements it. Here's an example of a valid `CompleteMultipartUpload` request from [S3 docs]:

```
POST /Key+?uploadId=UploadId HTTP/1.1
Host: Bucket.s3.amazonaws.com
x-amz-checksum-crc32: ChecksumCRC32
x-amz-checksum-crc32c: ChecksumCRC32C
x-amz-checksum-sha1: ChecksumSHA1
x-amz-checksum-sha256: ChecksumSHA256
x-amz-request-payer: RequestPayer
x-amz-expected-bucket-owner: ExpectedBucketOwner
x-amz-server-side-encryption-customer-algorithm: SSECustomerAlgorithm
x-amz-server-side-encryption-customer-key: SSECustomerKey
x-amz-server-side-encryption-customer-key-MD5: SSECustomerKeyMD5
<?xml version="1.0" encoding="UTF-8"?>
<CompleteMultipartUpload xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
   <Part>
      <ChecksumCRC32>string</ChecksumCRC32>
      <ChecksumCRC32C>string</ChecksumCRC32C>
      <ChecksumSHA1>string</ChecksumSHA1>
      <ChecksumSHA256>string</ChecksumSHA256>
      <ETag>string</ETag>
      <PartNumber>integer</PartNumber>
   </Part>
   ...
</CompleteMultipartUpload>
```

 [S3 docs go on to say](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html#API_CompleteMultipartUpload_RequestBody) that the `Part` array is not required, but it also says "If you do not supply a valid Part with your request, the service sends back an HTTP 400 response", which seems contradictory. In any case, if I construct the request like this:

```
let request = CompleteMultipartUploadRequest {
                    bucket: self.bucket.to_string(),
                    key: self.key.to_string(),
                    upload_id: self.upload_id.clone(),
                    multipart_upload: Some(CompletedMultipartUpload {
                        parts: None, // no parts, should be OK per AWS docs
                    }),
                    ..Default::default()
                }
```

The resulting request is this:

```
POST /fake-bucket/fake-key?uploadId=fake-upload-id
authorization: AWS4-HMAC-SHA256 Credential=ASIA5VMFCNSHIUNMH5VL/20220330/fake-region/s3/aws4_request, SignedHeaders=content-type;host;x-amz-content-sha256;x-amz-date;x-amz-security-token, Signature=590a269c01369f4476ca56e3df88f77c9fcb270cb463bd6098cf953d98134d7b
content-length: 65
content-type: application/octet-stream
host: 127.0.0.1:1234
x-amz-content-sha256: 6d82ee892d9a44cd4bea6319b348bb3eb36cd9e95628de19ffa0eb33a8161366
x-amz-date: 20220330T181325Z
x-amz-security-token: <snip>
user-agent: rusoto/0.47.0 rust/1.59.0 linux
<?xml version="1.0" encoding="utf-8"?><CompleteMultipartUpload />
```

And empirically, it's rejected by S3. Same goes if I set `CompleteMultipartUploadRequest.multipart_upload.parts` to an empty vec. I think maybe a request with a body like this would work:

```
<?xml version="1.0" encoding="utf-8"?>
<CompleteMultipartUpload>
    <Part />
</CompleteMultipartUpload>
```

... but I don't see a way to get Rusoto to construct it, and I'm not convinced it's worth the effort, since uploading an empty object isn't a use case we actually need to support.

Reply:
I think the simplest fix would probably be to upload an "empty" part. AFAICT nothing in the docs won't let this happen, and you'd be able to complete the upload in that case. This would be simple enough to do with minor changes to the existing code: have `upload_part` take a `bool` parameter that determines if this is the last chunk (similar to how GCS implementation's `upload_chunk` takes a parameter determining if this is the last chunk). Then, update the logic that short-circuits out of `upload_part` if `self.buffer.is_empty()` to upload an empty part if this is the last part & we have no completed parts so far.

As an alternate strategy, folks on the Internet report success doing a non-multipart upload in the empty-body case, too. But that's probably more code churn. It would save us a few network round-trips, though.

As Dave points out, we're already working around this bug elsewhere in the code, so fixing it would allow us to drop some workarounds! I didn't fix it when I ran into it previously because I wasn't sure how this would affect the collector. (we regularly encounter this issue when writing to the collector, as most `invalid_uuid` packet files are empty)

Reply:
Actually, thinking about the Collector issue -- given that most `invalid_uuid` packet files are empty & therefore currently not written at all, I think this PR would break us writing to the Collector (we'd regularly encounter NoCompletedParts errors, failing the overall write). I think we need to avoid this somehow.

Writing an empty object would work, _if_ the Collector implementation is such that it can accept empty-body `invalid_uuid` packet files as it currently accepts no object at all. I guess we'd need to confirm this with MITRE.

Reply:
Both the prod-us and prod-intl portal servers use GCS buckets, so I think this change won't affect communication with the collector concretely, though it would be a submarine bug in the event we ever wanted to set up a collector on AWS.

Reply:
Regarding that, I also think we should maintain the same semantics between AWS & GCP writers (since they are intended to be two different implementations of the same interface, and different implementations having surprising behavioral differences can lead to surprising bugs, as you note).

Reply:
Doing this unearthed a bug in how we handle errors from `ureq` in the GCS side. I think when we moved to `ureq` > 2.0, it started representing HTTP 4xx and 5xx as errors, so we were never noticing the HTTP 499 from canceling GCS uploads and instead treating them as errors. However we would only ever cancel when already in an error condition, so all this affected was which error gets logged.

Reply:
+1, not fixing the bug at all is probably the easiest path (if we can do so w/o un-fixing the bad behavior this PR is meant to fix). The new "you may not create 0-byte objects" documented semantics is kinda funky, but consistent with the pre-existing behavior.